File: dcp-solutions/2024-12-08/solution.c
```c
#include <stdbool.h>

// Definition for a binary tree node.
typedef struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

bool isValidBST(TreeNode* root) {
    return helper(root, INT_MIN, INT_MAX);
}

bool helper(TreeNode* node, long long minVal, long long maxVal) {
    if (node == NULL)
        return true;

    if (node->val <= minVal || node->val >= maxVal)
        return false;

    return helper(node->left, minVal, node->val) &&
           helper(node->right, node->val, maxVal);
}
```

File: dcp-solutions/2024-12-08/solution.py
```python
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isValidBST(root: TreeNode) -> bool:
    return helper(root, float('-inf'), float('inf'))

def helper(node: TreeNode, lower_bound: float, upper_bound: float) -> bool:
    if node is None:
        return True
    
    if not (lower_bound <= node.val <= upper_bound):
        return False
    
    return (helper(node.left, lower_bound, node.val) and 
            helper(node.right, node.val, upper_bound))
```